gtdynamics/factors/LinkPoseFactor.h

```diff
@@ -0,0 +1,109 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file  LinkPoseFactor.h
+ * @brief Factor relating a parent link and a child link given joint
+ * measurement.
+ * @author Varun Agrawal
+ */
+
+#pragma once
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+#include "gtdynamics/universal_robot/Joint.h"
+#include "gtdynamics/universal_robot/Link.h"
+
+namespace gtdynamics {
+
+/**
+ * @brief A 2-way factor to relate the parent and child links of a joint given
+ * the joint coordinate as a measurement.
+ *
+ * @tparam JOINT The type of joint which is being constrained.
+ */
+template <typename JOINT>
+class LinkPoseFactor
+    : public gtsam::NoiseModelFactor2<gtsam::Pose3, gtsam::Pose3> {
+ private:
+  using This = LinkPoseFactor;
+  using Base = gtsam::NoiseModelFactor2<gtsam::Pose3, gtsam::Pose3>;
+
+  JointConstSharedPtr joint_;
+  typename JOINT::JointCoordinate joint_coordinate_;
+  size_t k_;
+
+ public:
+  /**
+   * @brief Construct a new Link Pose Factor object
+   *
+   * @param wTp_key Key to the parent link.
+   * @param wTc_key Key to the child link.
+   * @param model The noise model for this factor.
+   * @param joint The joint connecting the links.
+   * @param joint_coordinate The coordinates of the joint motion.
+   * @param k The time index.
+   */
+  LinkPoseFactor(gtsam::Key wTp_key, gtsam::Key wTc_key,
+                 const gtsam::noiseModel::Base::shared_ptr& model,
+                 const JointConstSharedPtr joint,
+                 const typename JOINT::JointCoordinate& joint_coordinate,
+                 size_t k)
```

Comment:
`PoseFactor` avoids needing to pass in `k` by using `wTp_key.time()`.  That might help avoid situations where someone accidentally passes the wrong `k` and then gets a runtime error.

Reply:
Huh then I've been doing this all wrong so far. I should have been using this idiom all along.

Reply:
OKay there's a flaw in this approach in that I should be able to pass in a `gtsam::Key`. For example, a lot of my work with the IMU factors involves using a `gtsam::Key` to refer to the base link of a robot.

Reply:
gtd::DynamicsSymbol can be constructed from a key, so for example could do

`LinkPoseFactor(gtd::DynamicsSymbol(key1), gtd::DynamicsSymbol(key2), model, joint, 0.1)`

Or could overload the constructor

If `key1` and `key2` don't have semantically meaningful time indices to begin with then it kind of gets messed up anyway to connect a properly timed joint to a non-timed link?

I think it might be good practice for all links to be referred to by DynamicsSymbols rather than raw keys?